### PR TITLE
Simplify extra large domains in JDBC connectors by default

### PR DESCRIPTION
**Pull request by sopel39**
Supersedes: https://github.com/prestosql/presto/pull/5363

### CODE REVIEW

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/PredicatePushdownController.java [hunk 1]
```diff
@@ -14,15 +14,28 @@
...
+import static io.prestosql.plugin.jdbc.JdbcMetadataSessionProperties.getDomainCompactionThreshold;
 import static java.util.Objects.requireNonNull;
 
 public interface PredicatePushdownController
 {
-    PredicatePushdownController FULL_PUSHDOWN = (session, domain) -> new DomainPushdownResult(domain, Domain.all(domain.getType()));
-    PredicatePushdownController PUSHDOWN_AND_KEEP = (session, domain) -> new DomainPushdownResult(domain, domain);
-    PredicatePushdownController DISABLE_PUSHDOWN = (session, domain) -> new DomainPushdownResult(Domain.all(domain.getType()), domain);
+    PredicatePushdownController FULL_PUSHDOWN = (session, domain) -> {
+        if (getDomainSize(domain) > getDomainCompactionThreshold(session)) {
```
Comment (losipiuk) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/PredicatePushdownController.java:27]: Add to commit message that you actually enable simplification of pushed down domain in this comit

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClient.java [hunk 2]
```diff
@@ -126,8 +127,6 @@
 
     private static final int PRECISION_OF_UNSPECIFIED_NUMBER = 127;
 
-    private static final int ORACLE_MAX_LIST_EXPRESSIONS = 1000;
```
Comment (losipiuk) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClient.java:129]: With this you are changing the default value to 32. Why so?

**Review by losipiuk — COMMENTED** [review]
What is the motivation?

**Comment by sopel39** [comment]
> What is the motivation?

1. There was no compaction threshold for most connectors. Having `>1000` IN lists causes significant query slowdown. Even `>100` lists cause slowdowns. In other parts of code we use `32` and it seems line reasonable balance.
2. No need for special mechanism for Oracle and Sql server for compaction threshold. We hit limits for these connectors, but all connectors have similar limit. If not hard limit, then performance limit.
3. If somebody really needs to use larger lists, he still can do it per query basis (via session property)

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClient.java (hunk 2 above)
Comment (sopel39) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClient.java:129]: It provides best balance in general use. Large IN lists really cause significant perf regresions.

**Comment by losipiuk** [comment]
> > What is the motivation?
> 
> 1. There was no compaction threshold for most connectors. Having `>1000` IN lists cause significant query slowdown. Even `>100` lists cause regressions. In other parts of code we use `32` and it seems line reasonable balance.
> 2. No need for special mechanism for Oracle and Sql server for compaction threshold. We hit limits for these connectors, but all connectors have similar limit. If not hard limit, then performance limit.
> 3. If somebody really needs to use larger lists, he still can do it per query basis (via session property) 

Thanks. Can you put that in commit message. Or (IMO preferably) in code comment?

**Review by losipiuk — APPROVED** [review]

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataConfig.java [hunk 3]
```diff
@@ -17,10 +17,17 @@
...
 import io.airlift.configuration.LegacyConfig;
 
+import javax.validation.constraints.Min;
+
 public class JdbcMetadataConfig
 {
     private boolean allowDropTable;
     private boolean aggregationPushdownEnabled = true;
+    // Pushed domains are transformed into SQL IN lists in JDBC connectors.
+    // Too large IN lists cause significant performance regression.
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataConfig.java:27]: more precisely, to IN list, or sequence of OR'd range conditions

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataSessionProperties.java [hunk 4]
```diff
@@ -64,4 +70,17 @@ public static int getDomainCompactionThreshold(ConnectorSession session)
     {
         return session.getProperty(DOMAIN_COMPACTION_THRESHOLD, Integer.class);
     }
+
+    private static void validateDomainCompactionThreshold(int domainCompactionThreshold, Optional<Integer> maxDomainCompactionThreshold)
+    {
+        if (domainCompactionThreshold < 1) {
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "Domain compaction threshold must be greater than 0");
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataSessionProperties.java:77]: Quote the offending value

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataSessionProperties.java [hunk 5]
```diff
@@ -64,4 +70,17 @@ public static int getDomainCompactionThreshold(ConnectorSession session)
...
+        }
+
+        maxDomainCompactionThreshold.ifPresent(max -> {
+            if (domainCompactionThreshold > max) {
+                throw new PrestoException(INVALID_SESSION_PROPERTY, format("Domain compaction threshold cannot exceed %s", max));
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadataSessionProperties.java:82]: Quote the offending value

**Review by findepi — APPROVED** [review]

**Pull request merged**